### PR TITLE
perf(decode): single-pass MoE top-k + fused RoPE/KV-append (+3.2% decode)

### PR DESCRIPTION
**Pull request by fansilas**
## Summary

Two per-layer decode-graph nodes are latency-bound rather than bandwidth-bound, so on the replayed decode graph they cost wall-clock out of proportion to the work they do. This removes that overhead with two bit-exact changes (both default on, each independently toggleable):

**1. Single-pass MoE top-k (`SPARKINFER_ROUTER2`).** `moe_router` selected the top-8 of 128 experts with a single warp doing 8 serial arg-max passes. It now does one parallel sweep: one thread per expert counts how many experts outrank it (higher logit, or equal logit with a lower index — the exact tie-break the k-pass kernel used), which is its rank; experts with rank < top_k are the selection, written at slot == rank. Output order and softmax weights are bit-identical.

**2. Fused RoPE + KV-append (`SPARKINFER_ROPEKV`).** RoPE and the paged KV-cache append were two nodes. They fuse into one kernel that ropes Q in place, ropes K straight into the KV cache, and copies V into the cache, dropping the intermediate K round-trip. The roped Q/K are bit-identical to the standalone RoPE and the cached V is identical to the standalone append.

Distinct from the activation-quant/norm fusion and the Q/K-norm+RoPE fusion already in the tree: this targets the MoE routing and the KV-write nodes.

## Proof of speedup

- [x] Tested on **RTX 5090** (`sm_120`)

**Decode tok/s** (`bench/scripts/bench.sh`, Qwen3-30B-A3B Q4_K_M, n=128, bs=1):

| | decode tok/s |
|---|--:|
| before (both flags `=0`) | 384.5 |
| after (this PR) | 396.8 |

About **+3.2%** decode (mean of 6 alternating runs). Per-lever: single-pass top-k ~+1.95%, RoPE+KV-append fusion ~+1.06%, additive.

```text
run 1: BOTH_on=397.15  main=385.10
run 2: BOTH_on=397.38  main=383.76
run 3: BOTH_on=396.07  main=384.72
run 4: BOTH_on=396.79  main=385.15
run 5: BOTH_on=396.29  main=383.60
run 6: BOTH_on=396.86  main=384.76
```

## Accuracy (bit-identical to baseline)

`bench/scripts/accuracy.sh` vs llama.cpp, both flags on: token-match (top-1) 0.970, mean KL 0.1417 nats — equal to the baseline, as expected for bit-exact changes. `ctest`: 6/6 passed. `compute-sanitizer --tool memcheck`: 0 errors.

### CODE REVIEW

**Comment by ai-hpc** [comment]
## ✅ sparkinfer auto-eval — `1fafcc2`

| metric | value |
|---|---|
| **label** | `eval:S` |
| decode | 408.23 tok/s |
| vs prior frontier | 394.03 tok/s → +3.6% (+14.2) |
| **→ new frontier** | **408.23 tok/s** |
| correctness | top-1 97.0% · KL 0.1417 |

Verified speedup — **sets the new frontier to 408.23 tok/s** (was 394.03).

_RTX 5090 (sm_120) · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._

**Comment by ai-hpc** [comment]
Heads up — #83 just merged (frontier → **410.85 tok/s**), and this branch **conflicts with the new `main`** in `runtime/src/models/qwen35.cpp` (you both touch the decode loop). Please **rebase onto `main` and resolve** the conflict, then push — it'll be re-evaluated for its **marginal gain on top of #83** (your optimization looks independent, so it should still stack). Your verified result stands; it just needs to re-measure against the current frontier.

**Comment by fansilas** [comment]
Hi @ai-hpc Could you please review this again? Thanks!

**Comment by ai-hpc** [comment]
## ❌ sparkinfer auto-eval — `0f483c0`

| metric | value |
|---|---|
| **label** | `eval:REJECT` |
| decode | 0 tok/s |
| vs frontier | 419.84 tok/s → +0.0% (+0.0) |
| correctness | top-1 0.0% · KL 99 |

Failed the correctness gate: build failed (does not compile). Not a valid submission.

_RTX 5090 (sm_120) · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._

**Comment by ai-hpc** [comment]
Re-eval result: **`eval:REJECT` — does not compile.** Your rebase onto `main` (which now includes #83's residual-Q8_1 change in `runtime/src/models/qwen35.cpp`) left a build error — the conflict resolution didn't compile. The optimization itself is fine (it scored `S` before #83 landed); please **fix the merge so it builds**, then it'll re-evaluate for its marginal gain on the current frontier (419.84 tok/s baseline on this box). Full log: https://gittensor-ai-lab.github.io/sparkinfer-log/?run=0086-0f483c0

**Comment by fansilas** [comment]
I'm sorry, I'll fix it soon. Thanks for your patience!

**Comment by ai-hpc** [comment]
## ✅ sparkinfer auto-eval — `4e45ed9`

| metric | value |
|---|---|
| **label** | `eval:S` |
| decode | 454.41 tok/s |
| vs prior frontier | 436.21 tok/s → +4.2% (+18.2) |
| **→ new frontier** | **454.41 tok/s** |
| correctness | top-1 97.0% · KL 0.1417 |

Verified speedup — **sets the new frontier to 454.41 tok/s** (was 436.21).

_RTX 5090 (sm_120) · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._

**Comment by fansilas** [comment]
Thank you, @ai-hpc
